File: LeetCodeTasks/3013_divide-an-array-into-subarrays-with-minimum-cost-ii/solutions/java/solution.java
import java.util.Arrays;

class Solution {
    public long minimumCost(int[] nums, int k, int dist) {
        int n = nums.length;
        long[] dp = new long[n];
        Arrays.fill(dp, Integer.MAX_VALUE);
        
        for (int i = 0; i < k; i++) {
            long[] next = new long[n];
            long min = Long.MAX_VALUE;
            for (int j = i; j < n; j++) {
                if (j > 0) min = Math.min(min, dp[j - 1]);
                next[j] = Math.min(next[j], min + nums[j]);
                if (j - i <= dist) dp[j] = next[j];
            }
        }
        
        return dp[n - 1];
    }
}

File: LeetCodeTasks/3013_divide-an-array-into-subarrays-with-minimum-cost-ii/solutions/cpp/solution.cpp
#include <vector>
#include <algorithm>
#include <deque>

using namespace std;

class Solution {
public:
    long long minimumCost(vector<int>& nums, int k, int dist) {
        int n = nums.size();
        vector<long long> dp(n, INT_MAX);
        deque<int> dq;
        long long ans = INT_MAX;
        
        for (int i = 0; i < n; ++i) {
            if (!dq.empty())
                dp[i] = dp[dq.front()] + nums[i];
            while (!dq.empty() && dp[i] <= dp[dq.back()])
                dq.pop_back();
            dq.push_back(i);
            if (i - k + 1 >= 0 && !dq.empty() && dq.front() == i - k)
                dq.pop_front();
            if (i >= k - 1)
                ans = min(ans, dp[i]);
        }
        
        return ans;
    }
};
